File: src/lib/llm.ts
export interface LLMConfig {
  provider: "ollama" | "openai";
  baseUrl?: string;
  model: string;
  apiKey?: string;
}

export interface Message {
  role: "system" | "user" | "assistant";
  content: string;
}

export interface ChatCompletion {
  id: string;
  model: string;
  choices: Array<{
    message: Message;
    finish_reason: string;
  }>;
  usage: {
    prompt_tokens: number;
    completion_tokens: number;
    total_tokens: number;
  };
}

interface OllamaResponse {
  message?: { content?: string };
}

interface OpenAIResponse {
  choices?: Array<{ message?: { content?: string } }>;
}

export class LLMService {
  private config: LLMConfig;
  private abortController: AbortController | null = null;

  constructor(config: LLMConfig) {
    this.config = config;
  }

  static fromEnv(): LLMService {
    const ollamaUrl = process.env.OLLAMA_BASE_URL;
    const ollamaModel = process.env.OLLAMA_MODEL || "llama3.2";
    const openaiKey = process.env.OPENAI_API_KEY;
    const openaiModel = process.env.OPENAI_MODEL || "gpt-4o-mini";

    const provider: "ollama" | "openai" = openaiKey ? "openai" : "ollama";

    return new LLMService({
      provider,
      baseUrl: provider === "ollama" ? ollamaUrl : undefined,
      model: provider === "ollama" ? ollamaModel : openaiModel,
      apiKey: provider === "openai" ? openaiKey : undefined,
    });
  }

  async chat(messages: Message[]): Promise<string> {
    if (this.config.provider === "ollama") {
      return this.ollamaChat(messages);
    } else {
      return this.openaiChat(messages);
    }
  }

  async *streamChat(
    messages: Message[],
    onToken?: (token: string) => void,
  ): AsyncGenerator<string, void, unknown> {
    if (this.abortController) {
      throw new Error("Stream already in progress");
    }

    this.abortController = new AbortController();

    try {
      if (this.config.provider === "ollama") {
        yield* this.ollamaStreamChat(messages, onToken);
      } else {
        yield* this.openaiStreamChat(messages, onToken);
      }
    } finally {
      this.abortController = null;
    }
  }

  abortStream(): void {
    this.abortController?.abort();
  }

  private async *ollamaStreamChat(
    messages: Message[],
    onToken?: (token: string) => void,
  ): AsyncGenerator<string> {
    if (!this.config.baseUrl) {
      throw new Error("OLLAMA_BASE_URL is not configured");
    }

    const response = await fetch(`${this.config.baseUrl}/api/chat`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model: this.config.model,
        messages,
        stream: true,
      }),
      signal: this.abortController?.signal,
    });

    if (!response.ok) {
      throw new Error(`Ollama API error: ${response.statusText}`);
    }

    if (!response.body) {
      throw new Error("Response body is null");
    }
    const reader = response.body.getReader();
    const decoder = new TextDecoder();

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      const chunk = decoder.decode(value);
      const lines = chunk.split("\n").filter(Boolean);

      for (const line of lines) {
        try {
          const data = JSON.parse(line) as OllamaResponse;
          if (data.message?.content) {
            onToken?.(data.message.content);
            yield data.message.content;
          }
        } catch {
          // Skip invalid JSON
        }
      }
    }
  }

  private async *openaiStreamChat(
    messages: Message[],
    onToken?: (token: string) => void,
  ): AsyncGenerator<string> {
    if (!this.config.apiKey) {
      throw new Error("OPENAI_API_KEY is not configured");
    }

    const response = await fetch("https://api.openai.com/v1/chat/completions", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.config.apiKey}`,
      },
      body: JSON.stringify({
        model: this.config.model,
        messages,
        stream: true,
      }),
      signal: this.abortController?.signal,
    });

    if (!response.ok) {
      throw new Error(`OpenAI API error: ${response.statusText}`);
    }

    if (!response.body) {
      throw new Error("Response body is null");
    }
    const reader = response.body.getReader();
    const decoder = new TextDecoder();

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      const chunk = decoder.decode(value);
      const lines = chunk.split("\n").filter((line) => line.trim());

      for (const line of lines) {
        if (line.startsWith("data: ")) {
          const data = line.slice(6);
          if (data === "[DONE]") continue;

          try {
            const parsed = JSON.parse(data) as OpenAIResponse;
            const content = parsed.choices?.[0]?.message?.content;
            if (content) {
              onToken?.(content);
              yield content;
            }
          } catch {
            // Skip invalid JSON
          }
        }
      }
    }
  }

  private async ollamaChat(messages: Message[]): Promise<string> {
    if (!this.config.baseUrl) {
      throw new Error("OLLAMA_BASE_URL is not configured");
    }

    const response = await fetch(`${this.config.baseUrl}/api/chat`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model: this.config.model,
        messages,
        stream: false,
      }),
    });

    if (!response.ok) {
      throw new Error(`Ollama API error: ${response.statusText}`);
    }

    const data = (await response.json()) as OllamaResponse;
    return data.message?.content || "";
  }

  private async openaiChat(messages: Message[]): Promise<string> {
    if (!this.config.apiKey) {
      throw new Error("OPENAI_API_KEY is not configured");
    }

    const response = await fetch("https://api.openai.com/v1/chat/completions", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.config.apiKey}`,
      },
      body: JSON.stringify({
        model: this.config.model,
        messages,
        stream: false,
      }),
    });

    if (!response.ok) {
      throw new Error(`OpenAI API error: ${response.statusText}`);
    }

    const data = (await response.json()) as OpenAIResponse;
    return data.choices?.[0]?.message?.content || "";
  }
}

File: src/lib/llm.py
"""LLM service for Ollama and OpenAI."""

from __future__ import annotations

import json
from typing import Any, AsyncGenerator

import httpx

from models.data import LLMConfig, MessageRole


class LLMService:
    """Service for interacting with LLM APIs (Ollama/OpenAI)."""

    def __init__(self, config: LLMConfig) -> None:
        self.config = config
        self._client: httpx.AsyncClient | None = None

    @classmethod
    def from_env(cls) -> LLMService:
        """Create LLMService from environment variables."""
        import os

        ollama_url = os.environ.get("OLLAMA_BASE_URL", "http://localhost:11434")
        ollama_model = os.environ.get("OLLAMA_MODEL", "llama3.2")
        openai_key = os.environ.get("OPENAI_API_KEY")
        openai_model = os.environ.get("OPENAI_MODEL", "gpt-4o-mini")

        provider = "openai" if openai_key else "ollama"

        return cls(
            LLMConfig(
                provider=provider,
                base_url=ollama_url if provider == "ollama" else None,
                model=openai_model if provider == "openai" else ollama_model,
                api_key=openai_key,
            )
        )

    async def chat(self, messages: list[dict[str, str]]) -> str:
        """Send a chat message and get response."""
        if self.config.provider == "ollama":
            return await self._ollama_chat(messages)
        else:
            return await self._openai_chat(messages)

    async def stream_chat(
        self, messages: list[dict[str, str]]
    ) -> AsyncGenerator[str, None]:
        """Stream chat responses."""
        if self.config.provider == "ollama":
            async for token in self._ollama_stream(messages):
                yield token
        else:
            async for token in self._openai_stream(messages):
                yield token

    async def _ollama_chat(self, messages: list[dict[str, str]]) -> str:
        """Chat with Ollama API."""
        if not self.config.base_url:
            raise ValueError("OLLAMA_BASE_URL is not configured")

        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{self.config.base_url}/api/chat",
                json={
                    "model": self.config.model,
                    "messages": messages,
                    "stream": False,
                },
            )
            response.raise_for_status()
            data = response.json()
            return data.get("message", {}).get("content", "")

    async def _openai_chat(self, messages: list[dict[str, str]]) -> str:
        """Chat with OpenAI API."""
        if not self.config.api_key:
            raise ValueError("OPENAI_API_KEY is not configured")

        async with httpx.AsyncClient() as client:
            response = await client.post(
                "https://api.openai.com/v1/chat/completions",
                headers={
                    "Authorization": f"Bearer {self.config.api_key}",
                    "Content-Type": "application/json",
                },
                json={
                    "model": self.config.model,
                    "messages": messages,
                    "stream": False,
                },
            )
            response.raise_for_status()
            data = response.json()
            return data.get("choices", [{}])[0].get("message", {}).get("content", "")

    async def _ollama_stream(
        self, messages: list[dict[str, str]]
    ) -> AsyncGenerator[str, None]:
        """Stream from Ollama API."""
        if not self.config.base_url:
            raise ValueError("OLLAMA_BASE_URL is not configured")

        async with httpx.AsyncClient() as client:
            async with client.stream(
                "POST",
                f"{self.config.base_url}/api/chat",
                json={
                    "model": self.config.model,
                    "messages": messages,
                    "stream": True,
                },
            ) as response:
                response.raise_for_status()
                async for line in response.aiter_lines():
                    if not line:
                        continue
                    try:
                        data = json.loads(line)
                        content = data.get("message", {}).get("content", "")
                        if content:
                            yield content
                    except json.JSONDecodeError:
                        continue

    async def _openai_stream(
        self, messages: list[dict[str, str]]
    ) -> AsyncGenerator[str, None]:
        """Stream from OpenAI API."""
        if not self.config.api_key:
            raise ValueError("OPENAI_API_KEY is not configured")

        async with httpx.AsyncClient() as client:
            async with client.stream(
                "POST",
                "https://api.openai.com/v1/chat/completions",
                headers={
                    "Authorization": f"Bearer {self.config.api_key}",
                    "Content-Type": "application/json",
                },
                json={
                    "model": self.config.model,
                    "messages": messages,
                    "stream": True,
                },
            ) as response:
                response.raise_for_status()
                async for line in response.aiter_lines():
                    if not line or not line.startswith("data: "):
                        continue
                    data = line[6:]
                    if data == "[DONE]":
                        break
                    try:
                        parsed = json.loads(data)
                        content = (
                            parsed.get("choices", [{}])[0]
                            .get("delta", {})
                            .get("content")
                        )
                        if content:
                            yield content
                    except json.JSONDecodeError:
                        continue
